### PR TITLE
Use haskell-ci for CI

### PR DESCRIPTION
**Pull request by phadej**
- Jobs for GHC-7.0 .. GHC-8.10
- Revert ee8aabcc48b98e9fdd6f5068f73a867acd45193d (use return)
- Make `cabal check` happy
- CI still runs hlint, but not on the tests files
- Bump QuickCheck bounds to >=2.14 && <2.15

This is an alternative to #76, I'd prefer this.

As a bonus, also `appveyor.yml` using https://hub.zhox.com/posts/introducing-haskell-dev/

cc @haskell/core-libraries-committee

### CODE REVIEW

**Comment by ndmitchell** [comment]
This is massively more verbose, and implicitly means I'm now on the hook for maintaining quite a lot of shell script. What's the advantage? I don't particularly find testing more GHC's very valuable, since this is a package with a very low chance of being upgraded.

**Comment by cartazio** [comment]
It’s less work than you’d expect. The regeneration workflow is super easy
and I’m certainly happy to show you if need be.

**Comment by hvr** [comment]
> I don't particularly find testing more GHC's very valuable, since this is a package with a very low chance of being upgraded.

It's just like with seatbelts or airbags; it may be unlikely they gonna be used, but we still invest time & energy to put them in place because when those unlikely things would happen they'd protect us against major damage. Packages like `filepath` are at the bottom of the dep-graph and almost everyone depends on them directly or indirectly; and the bigger the rev-dep of a package is, the more important it is to ensure that no mistakes were made *before* publishing to Hackage, as any mistake has the potential to break a vast majority of Cabal users out there!

And going forward, there's a good chance we're going to see `filepath` upgrades happen a lot more likely when Cabal get's implemented a feature to conveniently control the general solver-preference-weight of sticking to preinstalled packages (there's even a TODO in the code about that... right now we can only control this per-package; but it's easy to add a knob to control the global policy). Once that's implemented your argument will have even less weight... :-)

**Comment by phadej** [comment]
I released the current state of `haskell-ci` to Hackage. The header now says:

```yaml
# This Travis job script has been generated by a script via
#
#   haskell-ci '--hlint' '--hlint-yaml=.hlint.yaml' 'filepath.cabal'
#
# To regenerate the script (for example after adjusting tested-with) run
#
#   haskell-ci regenerate
#
# For more information, see https://github.com/haskell-CI/haskell-ci
#
# version: 0.10
#
```

With `haskell-ci` `.travis.yml` is not meant to be edited manually. It is a "compilation result" of input data (`filepath.cabal` and command line arguments in this case), yet we try to keep it readable.

**Comment by phadej** [comment]
ping

**Comment by cartazio** [comment]
Any technical barriers to me pushing merge on this?

**Comment by ndmitchell** [comment]
I have been ill for the last few weeks, just recovering. I expect to take a look at this in the next week or so.

File: filepath.cabal [hunk 1]
```diff
@@ -48,20 +48,19 @@ library
...
-    ghc-options: -main-is Test
     hs-source-dirs: tests
     other-modules:
         TestGen
         TestUtil
     build-depends:
         filepath,
         base,
-        QuickCheck >= 2.7 && < 2.14
+        QuickCheck >= 2.14 && < 2.15
```
Comment (ndmitchell) [inline filepath.cabal:66]: Why is this a good thing? We demand people to upgrade QuickCheck and a whole pile of other dependencies for what benefit?

**Comment by ndmitchell** [comment]
The problem with this approach is that it requires me to learn two CI systems. I already have one I use for all my other projects, so if I switch filepath to a totally different system, I have to learn/debug/diagnose/follow two. I appreciate this new system is now supported by other people, which makes this less problematic than it would have been, and if I was starting afresh I might not bother implementing my own CI system, but given the current state, I don't see the benefits. Given that there hasn't yet been an issue with filepath in many years, and nearly all the newly tested versions are impossible to install on older GHC's, I don't see the benefit.

The fact that I don't follow `cabal check` is because I think the Cabal made an error changing the syntax for how to demand newer Cabal versions, so in my code I deliberately don't follow that. The other warning about non-portable `-main-is` has never made any sense to me. In my CI scripts I run `cabal check`, then parse out these two warnings and discard them explicitly.

Alternatively, if the core-libraries committee feels this is a useful point to standardise on, now might be a good time to switch maintainers? I haven't been particularly active on `filepath` in a long time (not that this is a bad thing, I don't think the library calls for vast changes!). My motivation for picking it up was that it had become orphaned and abandoned, and since I was the original author, still felt a responsibility. But I'd be happy to take a back seat if a general approach to maintaining all core-libraries as a set is desirable, as it may well be that having filepath "a bit different" from the others has greater costs to the CLC than the benefits of my maintainership. Thoughts?

(To add emotional context, since that's something that doesn't usually communicate well in text, I'm ambivalent - I would not feel disappointed, sad or otherwise upset if someone else wants to take the reins. I had a gentle plan to try and scale back my project load, so it might even be quite good timing.)

File: filepath.cabal (hunk 1 above)
Comment (chessai) [reply filepath.cabal:66]: I am also confused by this change.

Comment (phadej) [reply filepath.cabal:66]: This is a test dependency. `QuickCheck` is one of the lightest packages you can imagine, and there isn't even any `tasty-quickcheck` or similar package to recompile after it. QuickCheck-2.10 (released in 2017) made

> Random strings and characters now include Unicode characters 

change, which for example broke `bytestring` test-suite. Allowing older `QuickCheck` thus introduces more variation in what kind of tests are actually run. Take that to the limit, and using single `QuickCheck` major series should be enough.

Specially note, that e.g. Stackage is not upgrading `filepath`, so their setup is not running `filepath` tests. Neither does GHC run boot libraries test suites.

In other words, this change only affects `filepath` CI (which is fine), and direct developers (which there are a handful only).

Tell me a good reason to allow building `filepath` tests with QuickCheck-2.7?
